storm-core/src/jvm/backtype/storm/blobstore/KeyVersion.java

```diff
@@ -0,0 +1,125 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.TreeSet;
+import java.util.Map;
+import java.util.List;
+
+/**
+ * Class hands over the version of the key to be stored within the zookeeper
+ */
+public class KeyVersion {
+  private static final Logger LOG = LoggerFactory.getLogger(Utils.class);
+  private final String BLOBSTORE_SUBTREE="/blobstore";
+  private final String BLOBSTORE_KEY_COUNTER_SUBTREE="/blobstorekeycounter";
+  private String key;
+  private NimbusInfo nimbusInfo;
+
+  public KeyVersion(String key, NimbusInfo nimbusInfo) {
+    this.key = key;
+    this.nimbusInfo = nimbusInfo;
+  }
+
+  public int getKeyVersion(Map conf) {
+    TreeSet<Integer> versions = new TreeSet<Integer>();
+    CuratorFramework zkClient = Utils.createZKClient(conf);
+    try {
+      // Key has not been created yet and it is the first time it is being created
+      if(zkClient.checkExists().forPath(BLOBSTORE_SUBTREE + "/" + key) == null) {
+        zkClient.create().creatingParentsIfNeeded().withMode(CreateMode.PERSISTENT)
+                .withACL(ZooDefs.Ids.OPEN_ACL_UNSAFE).forPath(BLOBSTORE_KEY_COUNTER_SUBTREE + "/" + key + "/" + 1);
+        return 1;
+      }
+
+      // When all nimbodes go down and one or few of them come up
+      // Unfortunately there might not be an exact way to know which one contains the most updated blob
+      // if all go down which is unlikely. Hence there might be a need to update the blob if all go down
+      List<String> stateInfoList = zkClient.getChildren().forPath(BLOBSTORE_SUBTREE + "/" + key);
+      LOG.debug("stateInfoList {} stateInfoList {}", stateInfoList.size(), stateInfoList);
+      if(stateInfoList.isEmpty()) {
+        return getKeyVersionCounterValue(zkClient, key);
+      }
+
+      LOG.debug("stateInfoSize {}", stateInfoList.size());
+      // In all other cases check for the latest version on the nimbus and assign the version
+      // check if all are have same version, if not assign the highest version
+      for (String stateInfo:stateInfoList) {
+        versions.add(Integer.parseInt(Utils.normalizeVersionInfo(stateInfo)[1]));
+      }
+
+      int currentCounter = getKeyVersionCounterValue(zkClient, key);
+      // This condition returns version when a nimbus crashes and comes up
+      if (!checkIfStateContainsCurrentNimbusHost(stateInfoList, nimbusInfo) && !nimbusInfo.isLeader()) {
+        if (versions.last() < currentCounter) {
+          return currentCounter;
+        } else {
+          return currentCounter - 1;
+        }
+      }
+      // Condition checks for an update scenario
+      if (stateInfoList.size() >= 1 && versions.size() == 1) {
+        if (versions.first() < getKeyVersionCounterValue(zkClient, key)) {
+          incrementCounter(zkClient, key, currentCounter);
+          return currentCounter + 1;
+        } else {
+          incrementCounter(zkClient, key, currentCounter);
+          return versions.first() + 1;
```

Comment:
@d2r addressed all issues for all the above comments except

Utils.java:

1252     // Normalize state
1253     public static String[] normalizeVersionInfo(String nimbusKeyVersionInfo) {
1254       String[] normalizeVersionInfo = new String[2];
1255       int lastIndex = nimbusKeyVersionInfo.lastIndexOf("-");
1256       normalizeVersionInfo[0] = nimbusKeyVersionInfo.substring(0, lastIndex);
1257       normalizeVersionInfo[1] = nimbusKeyVersionInfo.substring(lastIndex + 1);
1258       return normalizeVersionInfo;
1259     }
We should make a class out of BlobVersion, with name and version as members. It's easier to understand than a String[] value. Will do in next request